The unique_host router filter can lose routes

smarterclayton opened the pull request:
The unique_host filter grew over time to add additional functionality and
scenarios. As part of that growth the guarantees for how claiming routes
evicted and restored other routes that shared the same host or path were
weakened and eventually broken. Two routes that conflict by host and path
created at the same time where one is deleted will result in a 50% chance
that the host remains claimed by the deleted route.

Refactor unique_host to use a separate data structure - a host index - to
track the routes that are assigned to each host. Split the logic for
deciding which routes for a given host to use into side-effect free
functions that can be individually tested. In the core unique_host code
guarantee that when routes cover other routes that deletions are propagated
down to lower level plugins and that no route is left behind.

Add much stronger tests on the core data structure to ensure the logic of
unique_host is not broken.

Identified while adding #18658. Only the last two commits are unique.

smarterclayton commented:
/test unit

smarterclayton commented:
/retest

smarterclayton commented:
/retest

smarterclayton commented:
/retest

smarterclayton commented:
/retest

smarterclayton commented:
/retest

smarterclayton commented:
/retest

smarterclayton commented:
/test gcp

smarterclayton commented:
/retest

smarterclayton commented:
Ok, this is ready for review.  The last few commits are cleaning up the tests.

knobunc commented:
@openshift/networking PTAL

smarterclayton commented:
/test gcp

smarterclayton commented:
Trying out a minor fix in the last commit that keeps the work queue from getting out of order - instead of requeueing with a delay and grabbing the next queue item immediately, sleep inside the work loop.  This can delay shutdown of the queue until the lease expires, but keeps queued items in roughly insertion order and avoids followers from simply skipping to the next item and competing with the leader.  The followers get locked out of trying work, so they fall further and further behind increasing the chance of the write lease being held.

knobunc commented:
Everything looks good to me too... I just don't want to say the magic word because your last commit still says WIP.

smarterclayton commented:
Yeah, I'm trying to tighten up the failure modes around conflict retries.

On Thu, Apr 19, 2018 at 3:54 PM, OpenShift CI Robot <
notifications@github.com> wrote:

> @smarterclayton <https://github.com/smarterclayton>: The following tests
> *failed*, say /retest to rerun them all:
> Test name Commit Details Rerun command
> ci/openshift-jenkins/cmd ab5e947
> <https://github.com/openshift/origin/commit/ab5e9471e9badfeebe9efdfec8feb9ca140582e6>
> link
> <https://openshift-gce-devel.appspot.com/build/origin-ci-test/pr-logs/pull/19175/test_pull_request_origin_cmd/12718/> /test
> cmd
> ci/openshift-jenkins/gcp ab5e947
> <https://github.com/openshift/origin/commit/ab5e9471e9badfeebe9efdfec8feb9ca140582e6>
> link
> <https://openshift-gce-devel.appspot.com/build/origin-ci-test/pr-logs/pull/19175/test_pull_request_origin_extended_conformance_gce/19335/> /test
> gcp
> ci/openshift-jenkins/unit 4f9079a
> <https://github.com/openshift/origin/commit/4f9079ab309ce972bf480f09df1f422275483387>
> link
> <https://openshift-gce-devel.appspot.com/build/origin-ci-test/pr-logs/pull/19175/test_pull_request_origin_unit/12221/> /test
> unit
>
> Full PR test history <https://openshift-gce-devel.appspot.com/pr/19175>. Your
> PR dashboard <https://openshift-gce-devel.appspot.com/pr/smarterclayton>.
> Please help us cut down on flakes by linking to
> <https://github.com/kubernetes/community/blob/master/contributors/devel/flaky-tests.md#filing-issues-for-flaky-tests>
> an open issue
> <https://github.com/openshift/origin/issues?q=is:issue+is:open> when you
> hit one in your PR.
>
> Instructions for interacting with me using PR comments are available here
> <https://git.k8s.io/community/contributors/devel/pull-requests.md>. If
> you have questions or suggestions related to my behavior, please file an
> issue against the kubernetes/test-infra
> <https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:>
> repository. I understand the commands that are listed here
> <https://go.k8s.io/bot-commands>.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/openshift/origin/pull/19175#issuecomment-382860625>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABG_p6aT7wC59Vua6Vayajk1qUx4ICaVks5tqOtdgaJpZM4TDKVN>
> .
>

smarterclayton commented:
/test gcp

smarterclayton commented:
Ok, I found and tracked down the bug that was resulting in the previous failures.  When I added the host_index I needed to make the behavior of receiving a route modification that doesn't change path or host be simpler - if the route is the same resource version as the one we have cached, it's a no-op, if it's a changed resource version and the route is currently active, we need to tell the unique_host plugin that the route is activated.  unique_host doesn't need to know the details, just whether to pass the update down.

With that, this PR should be correct and all known flakes addressed (i added tests for the update with no-op and update of an active route to the hostindex_test).  Ready for final review.
